components/binding-usage-controller/internal/controller/usagekind/controller.go

```diff
@@ -0,0 +1,198 @@
+		return
+	}
+	c.queue.Add(key)
+}
+
+// Run begins watching and syncing.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	go func() {
+		<-stopCh
+		c.queue.ShutDown()
+	}()
+
+	c.log.Infof("Starting service binding usage controller")
+	defer c.log.Infof("Shutting down service binding usage controller")
+
+	if !cache.WaitForCacheSync(stopCh, c.listerSynced) {
+		c.log.Error("Timeout occurred on waiting for caches to sync. Shutdown the controller.")
+		return
+	}
+
+	wait.Until(c.worker, time.Second, stopCh)
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	if c.testHookAsyncOpDone != nil {
+		defer c.testHookAsyncOpDone()
+	}
+
+	key, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.reconcile(key.(string))
+	switch {
+	case err == nil:
+		c.queue.Forget(key)
+
+	case c.queue.NumRequeues(key) < c.maxRetries:
+		c.log.Debugf("Error processing %q (will retry - it's %d of %d): %v", key, c.queue.NumRequeues(key), c.maxRetries, err)
+		c.queue.AddRateLimited(key)
+
+	default: // err != nil and too many retries
+		c.log.Errorf("Error processing %q (giving up - to many retires): %v", key, err)
+		c.queue.Forget(key)
+	}
+
+	return true
+}
+
+func (c *Controller) reconcile(name string) error {
+	usageKind, err := c.kindLister.Get(name)
+
+	switch {
+	case err == nil:
+	case apiErrors.IsNotFound(err):
+		// absence in store means watcher caught the deletion
+		c.log.Debugf("Starting deletion process of UsageKind %q", name)
+		if err := c.reconcileDelete(name); err != nil {
+			return errors.Wrapf(err, "while deleting UsageKind %q", name)
+		}
+		c.log.Debugf("UsageKind %q was successfully deleted", name)
+		return nil
+	default:
+		return errors.Wrapf(err, "while getting UsageKind %q", name)
+	}
+
+	c.log.Debugf("Starting reconcile UsageKind add process of %s", name)
+	defer c.log.Debugf("Reconcile UsageKind add process of %s completed", name)
+
+	if err := c.reconcileAddUpdate(usageKind); err != nil {
+		return errors.Wrapf(err, "while processing %s", name)
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcileDelete(name string) error {
+	c.log.Debugf("Unregistering supervisor for usage kind %s", name)
+	return c.kindContainer.Unregister(controller.Kind(name))
+}
+
+func (c *Controller) reconcileAddUpdate(kind *api.UsageKind) error {
+	c.log.Debugf("Registering supervisor for usage kind %s", kind.Name)
+	rip, err := newResourceInterfaceProvider(c.clientPool, schema.GroupVersionKind{
+		Version: kind.Spec.Resource.Version,
+		Group:   kind.Spec.Resource.Group,
+		Kind:    kind.Spec.Resource.Kind,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "while creating resource interface provider")
+	}
+
+	supervisor := controller.NewGenericSupervisor(rip,
+		newLabelManipulator(kind.Spec.LabelsPath),
+		c.log.WithField("service", "generic-supervisor"))
```

Comment:
can we extend "generic-supervisor" with the name of resource?

Reply:
added logger field "kind"